A71-xds-fallback.md

```diff
@@ -127,94 +52,160 @@ to [xDS Protocol Specification][resource-does-not-exist].
+gRPC servers using the xDS configuration will share the same XdsClient instance
+keyed with a dedicated well-known key value.
 
-In some gRPC implementations, the internal data structures do not allow for more
-than a single xDS server. The implementation needs to be updated to handle
-multiple servers, maintaining their fallback order.
+### Change xDS Bootstrap to handle multiple xDS configuration servers
 
-#### Add multi-client support to CSDS
+Current gRPC xDS implementations only use the first xDS configuration server
+listed in the bootstrap JSON document. Fallback implementation requires changes
+to use server listed and keep track of their ordering as it defines the
+relative priority of the different xDS config servers.
 
-See [A40] for details on CSDS.
+### Change XdsClient to support fallback within each xDS authority
 
-Add add a field to the ClientConfig message in the CSDS response to indicate
-which channel target the data is associated with.
+The fallback process is initiated if both of the following conditions hold:
 
-### Temporary environment variable protection
+* There is a connectivity failure on the ADS stream, as described in [A57]:
+    - Channel reports TRANSIENT_FAILURE.
+    - The ADS stream was closed before the first server response had been
+    received.
+* At least one watcher exists for a resource that is not cached.
 
-This option will be behind `GRPC_EXPERIMENTAL_XDS_FALLBACK`. If this variable
-is unset or is false, only one xDS server will be read from the bootstrap
-file. 
+This may happen either when setting up a new channel or if the xDS control
+plane becomes unavailable after a resource change notification was received.
+
+XdsClients will need to be changed to support multiple ADS connections for each
+authority. Once the fallback process begins, impacted XdsClient will establish
+a connection to the next xDS control plane listed in the bootstrap JSON. Then
+then XdsClient will subscribed to all watched resources on that server and will
+update the cache based on the received responses.
+
+Connecting to the lower-priority servers does not close gRPC connections to the
+higher-priority servers. XdsClient will still wait for xDS resources on the ADS
+stream. Once such resource is received, XdsClient will close connections to the
+lower-priority server(s) and will use the resources from the higher priority
+servers.
+
+### Potential issues
+
+#### Config tears
+
+Config tears happen when the client winds up using some combination
+of resources from the primary and fallback servers at the same time, even
+though that combination of resources was never validated to work together. In
+theory, this can cause correctness issues where we might send traffic to
+the wrong location or the wrong way, or it can cause RPCs to fail. Note
+that this can happen only when the primary and fallback server use the same
+resource names.
+
+This may also happen in non-fallback cases with the configs that span several
```

Comment:
The way this paragraph is written makes it seem like the recommendation for control planes to inline the RouteConfig in the LDS resource is specifically talking about config tears for the non-fallback case.  I think we want to split this into two paragraphs to avoid this implication.  Also, some of the wording below is a little unclear.

I suggest the following text for the rest of this subsection:

"""
Note that config tears can also happen in non-fallback cases with configs that span several xDS resources, and this is not something that the client can address in that more general case, so this is something control planes already need to be aware of.  We considered trying to address it for the fallback case by adding the xDS server name to the XdsClient watcher API, so that it became part of the key for the dependent watchers, thus ensuring that we never mix resources from the primary or fallback servers.  However, this would have required significant implementation work and would have introduced a seldom-used and therefore less well-exercised code-path that would likely have reduced reliability instead of increasing it.  And since this would not have helped for the more general case anyway, we decided not to pursue it.

Note that the main case where config tears  is likely to come up in practice is HTTP filter configs split across LDS and RDS. We recommend that control planes avoid that problem by inlining the RouteConfiguration into the LDS resource.
"""